### ISSUE TITLE
storage: problematic dependency resolution with `stats`

### ISSUE THREAD

**Issue by StephenWithPH** [googleapis/google-cloud-go#11283]
Having upgraded from:

`cloud.google.com/go/storage v1.46.0` -> `cloud.google.com/go/storage v1.48.0`, I get the following error:

```text
<redacted>/go/pkg/mod/cloud.google.com/go/storage@v1.48.0/storage.go:264:25: undefined: stats.NewMetrics
```

This corresponds to:

https://github.com/googleapis/google-cloud-go/blob/d95c86d30f45d6d54b438579dbca20e699e51daa/storage/storage.go#L264

... and I believe these two imports are in conflict:

https://github.com/googleapis/google-cloud-go/blob/d95c86d30f45d6d54b438579dbca20e699e51daa/storage/storage.go#L56-L57

https://github.com/googleapis/google-cloud-go/blob/d95c86d30f45d6d54b438579dbca20e699e51daa/storage/go.mod#L26-L27

I suspect that somehow, this [`NewMetrics`](https://pkg.go.dev/google.golang.org/grpc@v1.67.2/experimental/stats#NewMetrics) is getting shadowed by https://pkg.go.dev/google.golang.org/grpc/stats/opentelemetry@v0.0.0-20241028142157-ada6787961b3 .

For a brief period, I was getting ambiguous import errors pointing at these two packages with `go mod tidy`. I'm no longer getting _those_ errors, but I still fail to compile.

**Comment by davidh16** [comment]
same issue here

**Comment by SadriShehu** [comment]
Facing the same issue as well.

**Comment by fasmat** [comment]
The problem is that `google.golang.org/grpc/stats/opentelemetry` is not its own module any more. This was changed between v1.68.1 and v1.69.0 of `google.golang.org/grpc`: https://github.com/grpc/grpc-go/pull/7759

Upgrading to the latest `google.golang.org/grpc` package will now cause the same package to be imported from two different sources. Upgrading `storage` to use the latest `google.golang.org/grpc` and dropping the `google.golang.org/grpc/stats/opentelemetry` dependency should fix the issue.

**Comment by fasmat** [comment]
I created a PR to update the GRPC dependency in all modules: https://github.com/googleapis/google-cloud-go/pull/11284

**Comment by nikolaydubina** [comment]
I mean, that change is huge. not sure can be accepted

need someone from google to fix this

**Comment by nikolaydubina** [comment]
ok, updating everything and keeping `cloud.google.com/go/storage v1.46.0` worked for me. will check back on this once google replies.

**Comment by dfawley** [comment]
I believe the only (?) issue here is that grpc-go moved some of the symbols that cloud storage was relying upon from our `experimental` directory into a stable location, without leaving behind type aliases.  If so, that's our mistake and we will fix it ASAP.  PR https://github.com/grpc/grpc-go/pull/7929 should address it, and we will do a patch release with it ASAP.

**Comment by fasmat** [comment]
@dfawley I'm not sure this solves the issue. In `v1.69.0` `google.golang.org/grpc/stats/opentelemetry` is a package of the `google.golang.org/grpc` module. In previous versions of `google.golang.org/grpc` it is not. So you still have the problem where if you use `cloud.google.com/go/storage` and `google.golang.org/grpc` as direct dependencies you cannot upgrade the later to `v1.69.0` without breaking the dependency tree.

**Comment by dfawley** [comment]
@fasmat Thanks for confirming that the change above will not fully fix things.  [This document](https://go.dev/wiki/Modules#is-it-possible-to-remove-a-module-from-a-multi-module-repository) seems to indicate that such an operation is possible and valid, but it's missing details on how to perform the removal -- it just refers to the "adding a module" instructions and how to actually do it correctly is left as an exercise for the reader.  In particular, _splitting_ a module requires adding a dependency from the new module to a new version of the existing module, but when _joining_ modules, I can't add a dependency to the _removed_ module, since no module exists there anymore.

So for now I'm not sure how to _properly_ remove a module and move its packages to its parent module.  (If it is possible as the wiki suggests.)  I'll keep working on this.

**Comment by tritone** [comment]
Hey all, apologies for the issue here, we are working to fix this ASAP.

There is a workaround that should unblock compilation and `go get` calls, which is to exclude the latest release of grpc-go from your go.mod.

`exclude google.golang.org/grpc v1.69.0`

Let me know if this resolves this issue.

**Comment by gannett-ggreer** [comment]
@tritone 
> `exclude google.golang.org/grpc v1.69.0`
> 
> Let me know if this resolves this issue.

It avoids the dependency resolution failing at least, for the repositories I've tried.

The non-manual way is `go mod edit --exclude=google.golang.org/grpc@v1.69.0`.

**Comment by StevenACoffman** [comment]
@tritone @zasweq Is it possible to just add the `go.mod` back for `stats/opentelemetry` from #7759 and cut a point release until `storage` is ready to upgrade to Go 1.22 ?

**Comment by dfawley** [comment]
>  it possible to just add the go.mod back for `stats/opentelemetry`

My plan is to do the opposite.  Pull the module _into_ grpc-go v1.67.x and v1.68.x, so it's like it never existed.  Then googleapis can upgrade its version to those and everything should work moving forward, too.

I was hoping we could actually find a way to do this that worked automagically, just like how you can split modules without any coordination.  But no such approach seems possible, and https://go.dev/wiki/Modules#is-it-possible-to-remove-a-module-from-a-multi-module-repository should really be updated to say "no, not really" instead of "yes, just do the opposite of the above".

**Comment by StevenACoffman** [comment]
@dfawley You are probably already aware, but I wanted to get confirmation that I understand the plan for my colleagues.

+[grpc-go v1.67.2 uses Go 1.21](https://github.com/grpc/grpc-go/blob/b0e0bf0394b8caaafacf06e271d1f65406268e41/go.mod#L3)
+ [grpc-go v1.68.x uses Go 1.22.x](https://github.com/grpc/grpc-go/blob/v1.68.0/go.mod#L3)

However: 
+ [cloud.google.com/go/storage uses Go 1.21](https://github.com/googleapis/google-cloud-go/blob/main/storage/go.mod#L3) and cannot update to a newer version of Go until January and so cannot use any version of grpc-go  v1.68.x

So after you have merged #7935 and released grpc-go v1.67.3 , cloud.google.com/go/storage will need to update to that (not v1.68.x) and then cut a new release (maybe v1.50.0 since v1.49.0 release already has a PR for it and appears to be in progress)

In the meantime, `go mod edit --exclude=google.golang.org/grpc@v1.69.0` is the only workaround.

**Comment by codyoss** [comment]
@StevenACoffman I believe you summarized this well, yes.

**Comment by dfawley** [comment]
I have released grpc-go v1.67.3 that pulls in the previously-experimental module that cloud storage was using.

@StevenACoffman,
>In the meantime, `go mod edit --exclude=google.golang.org/grpc@v1.69.0` is the only workaround.

Just to clarify this a bit: after `cloud.google.com/**` upgrades to grpc-go v1.67.3, _you_ should be able to upgrade to grpc-go v1.69.x, since you (presumably) don't have the Go 1.21 restriction.  There should be no ambiguous import after that.

**Comment by tritone** [comment]
Okay, the latest release seems to fix the issue in my repro.

* `go get -u google.golang.org/grpc` (can upgrade to latest release v1.69.2 or v1.67.3 for compatibility with go 1.21)
* `go get -u cloud.google.com/go/storage`

You may also have to remove `google.golang.org/grpc/stats/opentelemetry` as an indirect dependency.

Let me know if this resolves the issue for you all or if there are still any problems.

Apologies for the inconvenience here.

**Comment by StevenACoffman** [comment]
BTW, I had to `exclude google.golang.org/grpc/stats/opentelemetry v0.0.0-20240907200651-3ffb98b2c93a`:
```
go mod edit --exclude=google.golang.org/grpc/stats/opentelemetry@v0.0.0-20240907200651-3ffb98b2c93a
go get -u google.golang.org/grpc
go get -u cloud.google.com/go/storage
```

**Comment by tritone** [comment]
> BTW, I had to `exclude google.golang.org/grpc/stats/opentelemetry v0.0.0-20240907200651-3ffb98b2c93a`:
> 
> ```
> go mod edit --exclude=google.golang.org/grpc/stats/opentelemetry@v0.0.0-20240907200651-3ffb98b2c93a
> go get -u google.golang.org/grpc
> go get -u cloud.google.com/go/storage
> ```

Yes, or manually deleting the indirect dep on google.golang.org/grpc/stats/opentelemetry from the go.mod should also do the trick.

**Comment by dfawley** [comment]
> Yes, or manually deleting the indirect dep on google.golang.org/grpc/stats/opentelemetry from the go.mod should also do the trick.

I _believe_ `go mod tidy` should take care of this for you, no?

**Comment by tritone** [comment]
> > Yes, or manually deleting the indirect dep on google.golang.org/grpc/stats/opentelemetry from the go.mod should also do the trick.
> 
> I _believe_ `go mod tidy` should take care of this for you, no?

Unfortunately in my repro, `go mod tidy` also didn't work until the indirect dep was manually removed.

**Comment by tritone** [comment]
> Okay, the latest release (storage v1.49.0) seems to fix the issue in my repro.
> 
> * `go get -u google.golang.org/grpc` (can upgrade to latest release v1.69.2 or v1.67.3 for compatibility with go 1.21)
> * `go get -u cloud.google.com/go/storage`
> 
> You may also have to remove `google.golang.org/grpc/stats/opentelemetry` as an indirect dependency.
> 
> Let me know if this resolves the issue for you all or if there are still any problems.
> 
> Apologies for the inconvenience here.

I haven't heard from anyone still experiencing this issue, so I'm going to close this out. Thanks for your patience all.

**Comment by mathieu-lemay** [comment]
You have to be kidding me. While this issue indeed seems to be fixed, [github.com/envoyproxy/go-control-plane](https://github.com/envoyproxy/go-control-plane) decided to join the party!

```
$ go get -u cloud.google.com/go/storage
go: cloud.google.com/go/storage imports
        google.golang.org/grpc/xds/googledirectpath imports
        google.golang.org/grpc/internal/xds/bootstrap imports
        github.com/envoyproxy/go-control-plane/envoy/config/core/v3: ambiguous import: found package github.com/envoyproxy/go-control-plane/envoy/config/core/v3 in multiple modules:
        github.com/envoyproxy/go-control-plane v0.13.1 (/Users/6004148/.local/share/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.13.1/envoy/config/core/v3)
        github.com/envoyproxy/go-control-plane/envoy v1.32.2 (/Users/6004148/.local/share/go/pkg/mod/github.com/envoyproxy/go-control-plane/envoy@v1.32.2/config/core/v3)
```

**Comment by dfawley** [comment]
Please file a bug in go-control-plane.  That issue is unrelated.

https://github.com/envoyproxy/go-control-plane/pull/714 probably didn't follow the right steps for _splitting_ modules:  

https://go.dev/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository

**Comment by tritone** [comment]
Ugh. I'll open a new issue here as well to track resolution, since it blocks updates for storage as well as other clients which have this as an indirect dependency (spanner, bigtable).

**Comment by tritone** [comment]
Follow along here: #11344

**Comment by AndreasBergmeier6176** [comment]
For us what fixed the error was:

> exclude google.golang.org/grpc/stats/opentelemetry v0.0.0-20240907200651-3ffb98b2c93a
